### ISSUE TITLE
CORS error accessing external api

### ISSUE THREAD

**Issue by cusmith1** [HelloZeroNet/ZeroNet#1045]
We are exploring options for negating the effects of DDoS attacks in enterprise web applications (with large backends). Currently we are considering using ZeroNet to serve as a decentralized frontend for the site while using Blockstack (blockstack.org) to manage decentralized control and encryption of user data (backend).

The zeronet zite however, doesn't allow the cross origin connection to the Blockstack browser for  authentication or data access. Errors as follows:

> Origin http://localhost:8888 is not allowed by Access-Control-Allow-Origin.
> Failed to load resource: Origin http://localhost:8888 is not allowed by Access-Control-Allow-Origin
> Fetch API cannot load http:127.0.0.1:43110/manifest.json. Origin http://localhost:8888 is not allowed by Access-Control-Allow-Origin.

I already attempted to use the zeronet CORS plugin to allow localhost:8888 which didn't work. Is there a special case for non-zeronet sites? or is there no way to use API external to zeronet?

**Comment by mkg20001** [comment]
You need to enable cors on blockstack. ZeroNet only enables cors for zeronet zites

**Comment by cusmith1** [comment]
Cors is enabled on blockstack. It is the zeronet side that is blocking the request.

**Comment by HelloZeroNet** [comment]
You have to use the fileGet api call instead of fetch api to load files

**Comment by cusmith1** [comment]
<img width="1395" alt="screen shot 2017-07-25 at 2 42 34 pm" src="https://user-images.githubusercontent.com/17167947/28588317-2fbf740e-7148-11e7-8dc1-28aa2e170ac8.png">
<img width="591" alt="screen shot 2017-07-25 at 2 43 32 pm" src="https://user-images.githubusercontent.com/17167947/28588324-32a2dfa8-7148-11e7-97eb-a1025acf220e.png">

The site supposedly has permissions for localhost:8888 where the Blockstack browser is but I still get cross origin errors.

Also can you point me to an explanation of how to use fileGet instead?

**Comment by HelloZeroNet** [comment]
Here is a simple example on API usage:
https://github.com/HelloZeroNet/ZeroHello/blob/master/template-new/index.html
The command you need: `page.cmd("fileGet", "content.json", function(data) { console.log(data) })`

**Comment by mkg20001** [comment]
@HelloZeroNet It's about reading one zite from another host. (example localhost:8080 -> localhost:43110/1Hello/)

**Comment by HelloZeroNet** [comment]
@mkg20001 It won't work in our situation, since the host is same for all sites.

**Comment by mkg20001** [comment]
@HelloZeroNet 
> Origin http://localhost:8888 is not allowed by Access-Control-Allow-Origin.
Failed to load resource: Origin http://localhost:8888 is not allowed by Access-Control-Allow-Origin
Fetch API cannot load http:127.0.0.1:43110/manifest.json. Origin http://localhost:8888 is not allowed by Access-Control-Allow-Origin.

I don't want to start a fight but those messages clearly indicate that ZeroNet is the one being accessed by/from the Blockstack host which does not work because ZeroNet does not set the right cors headers for requests from other software/hosts

**Comment by HelloZeroNet** [comment]
You can be right, I'm not sure what's that manifest.json file or what @cusmith1 tries to do exactly.

**Comment by grez911** [comment]
@cusmith1 Please, participate.

**Comment by cusmith1** [comment]
Apologies, I got sidetracked by another project and lost track of this.

The blockstack system is set up to use the Fetch API so that was part of the issue. ZeroNet was blocking that despite the CORS setup of the requests as it is necessary to use the FileGet method @HelloZeroNet described above. I was able to set it up with FileGet on the client side which removed most of the errors, but the base blockstack browser still requires Fetch in some spots resulting in a failed URI request on blockstacks side.

We're now looking at altering the base blockstack code to make it function with zeronet's API but that's for us to figure out with their side. Thank you all for giving your thoughts.
